docs: add extension flakiness patterns Cursor rule (#40637)
MetaMask/metamask-extension: .cursor/rules/extension-flakiness-patterns/RULE.md

    .withPreferencesController({ useNonceField: true })
    .build(),
  async ({ driver }) => {
    await login(driver);
  ```

  This pattern applies broadly — use `.withSelectedNetwork(NETWORK_CLIENT_ID.MAINNET)` instead of switching network via UI, `withTokensControllerERC20()` instead of importing tokens through modals, and pre-deploy contracts via fixture (`smartContract: [...]`) instead of deploying through the dapp UI.

- **Avoid unnecessary browser refreshes — they can cause unexpected navigation (e.g., landing on a Confirmation screen for unapproved transactions)**
  ❌ Incorrect:

  ```javascript
  await driver.switchToWindowWithTitle(WINDOW_TITLES.ExtensionInFullScreenView);
  await driver.executeScript(`window.location.reload()`);
  await driver.switchToWindowWithTitle(WINDOW_TITLES.ExtensionInFullScreenView);
  ```

  ✅ Correct:

  ```javascript
  await driver.switchToWindowWithTitle(WINDOW_TITLES.ExtensionInFullScreenView);
  await homePage.goToActivityList();
  ```

- **No hardcoded delays — use wait-for-conditions instead**
  ❌ Incorrect:

  ```javascript
  await driver.delay(5000);
  await driver.clickElement(selector);
  ```

  ✅ Correct:

  ```javascript

>   await driver.waitForSelector(selector);

Reviewer: this code snippet could lead to think that before clickElement we need to waitForSelector for that element.
That is implicit in the clickElement.

We could change the selector, so then it's clear that the 2 selectors are different:

```suggestion
  await driver.waitForSelector(title);
```

3. **Using `element.click()` instead of `driver.clickElement()`** — native `.click()` has no stale-element retry. Always use the driver wrapper. See [Not using driver methods](#not-using-driver-methods).

4. **Going to live sites** instead of using mocks — tests should never depend on external services. See [Missing or Incorrect Use of Mocks](#missing-or-incorrect-use-of-mocks).

5. **Adding `driver.delay()` instead of waiting for conditions** — use `waitForSelector`, `clickElementAndWaitToDisappear`, or `waitUntil` instead of arbitrary delays. See [Confirmation Popups / Modals](#confirmation-popups--modals).

6. **Importing from another `.spec` file** — this causes that spec's tests to run too, leading to timeouts. Move shared functions to helper files. See [Actions that Take Time](#actions-that-take-time).

7. **Looking for transient elements** (e.g., "Pending" status) — skip transient states and wait for the final state. See [Race Conditions with Assertions](#race-conditions-with-assertions-within-the-test-body-steps).

---

## E2E Flakiness Categories

### Race Conditions on Driver/Helpers Functions

- **Click Element with stale element**
  ❌ Incorrect:

  ```javascript

>   async clickElement(rawLocator) {

Reviewer: I would change this title, as it seems to imply that we recommend to write and modify our generic methods from the driver (we shouldn't do this often).
This recommendation should go for page object methods, so we could write some more representative name. Ie.

```suggestion
  async clickCloseButton(rawLocator) {
```

  await driver.clickElement(this.addAccountButton);
  ```

- **In the onboarding flow, clicking an element when it's moving causes the click to take no effect. Added a new driver method to wait until the element is not moving**
  ❌ Incorrect:

  ```javascript
  await driver.clickElement(this.categoryBackButton);
  await driver.delay(regularDelayMs);
  await driver.clickElement(this.privacySettingsBackButton);
  ```

  ✅ Correct:

  ```javascript
  await driver.clickElement(this.categoryBackButton);
  // Wait until carousel stops animating — clicking while moving has no effect
  await driver.waitForElementToStopMoving(this.privacySettingsBackButton);
  await driver.clickElement(this.privacySettingsBackButton);
  ```

- **In the carousel spec, looking for an element and then using `.click` — a re-render in between made the element stale. Should use the custom `clickElement` driver method**
  ❌ Incorrect:
  ```javascript
  const dots = await driver.findElements(this.carouselDot);
  await dots[i].click();
  ```
  ✅ Correct:
  ```javascript
  await driver.clickElement(`[aria-label="slide item ${i}"]`);
  ```

---

### Actions that Take Time

> > **General rule:** When an action takes time (API calls, file I/O, state updates, extension restarts), wait for the expected outcome using `driver.wait`, `waitUntil`, or `waitForSelector` instead of assuming it's complete. Use generous timeouts — 3 seconds is often not enough.

Reviewer: ```suggestion
> **General rule:** When an action takes time (API calls, file I/O, state updates, extension restarts), wait for the expected outcome using `driver.wait`, `waitUntil`, or `waitForSelector` instead of assuming it's complete and custom timeouts.```
